add VsockPacket implementation

lauralt opened the pull request:
This PR is adding an implementation of the vsock packet, based on the design proposal from https://github.com/rust-vmm/vm-virtio/issues/128. The documentation is in a decent shape already, but I will open a separate PR with some improvements to it and also code examples.
Open questions:
1. Do we want to have an upper limit for the data array size? I didn't add one yet, cloud-hypervisor and firecracker are using [the one defined in the linux kernel](https://elixir.bootlin.com/linux/v4.9.12/source/include/linux/virtio_vsock.h#L14). Do we assume this one to be a good enough limit for a vsock driver?
2. On the TX path, should we go with the `len` from the header for the data array size or do we keep the one from the data descriptor? I'll think a bit more about this.

lauralt commented:
@jiangliu @sboeuf Can you also have a look at this PR? Thank you!

andreeaflorescu commented:
> 1. Do we want to have an upper limit for the data array size? I didn't add one yet, cloud-hypervisor and firecracker are using [the one defined in the linux kernel](https://elixir.bootlin.com/linux/v4.9.12/source/include/linux/virtio_vsock.h#L14). Do we assume this one to be a good enough limit for a vsock driver?

Who is filling the data array? It is a good practice to set a size, using the same defined as in the Linux Kernel sounds like a good plan.

> 2. On the TX path, should we go with the `len` from the header for the data array size or do we keep the one from the data descriptor? I'll think a bit more about this.

I'll take a look at the code as well and come back to this.

lauralt commented:
> > 1. Do we want to have an upper limit for the data array size? I didn't add one yet, cloud-hypervisor and firecracker are using [the one defined in the linux kernel](https://elixir.bootlin.com/linux/v4.9.12/source/include/linux/virtio_vsock.h#L14). Do we assume this one to be a good enough limit for a vsock driver?
> 
> Who is filling the data array? It is a good practice to set a size, using the same defined as in the Linux Kernel sounds like a good plan.

It is filled by the driver on the path where it would make sense to check for this limit. @harshanavkis @stsquad Is this limit ok for you too? I've noticed you don't check for one in the vhost-user-vsock packet implementation.

stsquad commented:
> > > 1. Do we want to have an upper limit for the data array size? I didn't add one yet, cloud-hypervisor and firecracker are using [the one defined in the linux kernel](https://elixir.bootlin.com/linux/v4.9.12/source/include/linux/virtio_vsock.h#L14). Do we assume this one to be a good enough limit for a vsock driver?
> > Who is filling the data array? It is a good practice to set a size, using the same defined as in the Linux Kernel sounds like a good plan.
> It is filled by the driver on the path where it would make sense to check for this limit. @harshanavkis @stsquad Is this limit ok for you too? I've noticed you don't check for one in the vhost-user-vsock packet implementation.

AFAICT it's a fairly arbitrary limit from the Linux kernel which isn't part of the spec and my general feeling is not to embed specific implementation limitations into the back-end. However given there aren't any alternate vsock front-ends it's not the end of the world if it has a benefit to the code - does having a limit make things easier for the rust compiler?

andreeaflorescu reviewed (commented):
Part 1

harshanavkis commented:
> > > > 1. Do we want to have an upper limit for the data array size? I didn't add one yet, cloud-hypervisor and firecracker are using [the one defined in the linux kernel](https://elixir.bootlin.com/linux/v4.9.12/source/include/linux/virtio_vsock.h#L14). Do we assume this one to be a good enough limit for a vsock driver?
> > > >    Who is filling the data array? It is a good practice to set a size, using the same defined as in the Linux Kernel sounds like a good plan.
> > > >    It is filled by the driver on the path where it would make sense to check for this limit. @harshanavkis @stsquad Is this limit ok for you too? I've noticed you don't check for one in the vhost-user-vsock packet implementation.
> 
> AFAICT it's a fairly arbitrary limit from the Linux kernel which isn't part of the spec and my general feeling is not to embed specific implementation limitations into the back-end. However given there aren't any alternate vsock front-ends it's not the end of the world if it has a benefit to the code - does having a limit make things easier for the rust compiler?

Yes I didn't specify a data size in the packet implementation for the device because I think 64KiB is the maximum transfer size possible and the packet at runtime may be of a smaller data size because of flow control for SOCK_STREAM type vsock sockets. As far as I know [SEQ_PACKETS](https://lwn.net/Articles/856826/) doesn't enforce a max size of the data packet. And we will need to add it to the device in the future as well.

lauralt commented on crates/devices/virtio-vsock/src/packet.rs:
````diff
@@ -0,0 +1,386 @@
+// The vsock packet header is defined by the following structure:
+//
+// ```
+// struct virtio_vsock_hdr {
````
I'll give it a shot.

gsserge commented on crates/devices/virtio-vsock/src/packet.rs:
```diff
@@ -0,0 +1,934 @@
+        };
+
+        // If the `len` field of the header is zero, then the packet doesn't have a `data` element.
+        if pkt.len() == 0 {
```
I don't have a very strong opinion about this, but maybe it's worth it to implement `.is_empty()` and get rid of `#[allow(clippy::len_without_is_empty)]` if we are doing the empty check here.

lauralt replied:
Good idea, I'll add it.

gsserge commented on crates/devices/virtio-vsock/src/packet.rs:
```diff
@@ -0,0 +1,934 @@
+
+        Ok(Self {
+            header,
+            header_inner: Default::default(),
```
Why are we using the default values for `header_inner` instead of initializing it based on `header` similar to `from_tx_virtq_chain `?

lauralt replied:
On the RX path there is no data in the packet's header and data buffers. The driver is providing empty buffers to be later filled by the device.

lauralt commented:
> > > > 1. Do we want to have an upper limit for the data array size? I didn't add one yet, cloud-hypervisor and firecracker are using [the one defined in the linux kernel](https://elixir.bootlin.com/linux/v4.9.12/source/include/linux/virtio_vsock.h#L14). Do we assume this one to be a good enough limit for a vsock driver?
> > > >    Who is filling the data array? It is a good practice to set a size, using the same defined as in the Linux Kernel sounds like a good plan.
> > > >    It is filled by the driver on the path where it would make sense to check for this limit. @harshanavkis @stsquad Is this limit ok for you too? I've noticed you don't check for one in the vhost-user-vsock packet implementation.
> 
> AFAICT it's a fairly arbitrary limit from the Linux kernel which isn't part of the spec and my general feeling is not to embed specific implementation limitations into the back-end. However given there aren't any alternate vsock front-ends it's not the end of the world if it has a benefit to the code - does having a limit make things easier for the rust compiler?

The benefit wouldn't be to the code, but from what I know it's a good practice to have some sort of limit for arrays, such that the device implementation doesn't allow a driver to use super huge chunks of memory, and potentially fill up the entire memory. The same rationale was behind [this issue](https://github.com/rust-vmm/vm-virtio/issues/114) as well.

lauralt commented:
@andreeaflorescu @gsserge I addressed your comments.

Regarding the max vsock packet data size discussion, I opened [this thread](https://lists.oasis-open.org/archives/virtio-comment/202202/msg00053.html) in virtio-spec; the proposal from there would be to have a new field in the vsock configuration space for that. Since this will involve changes that are going to take some time, I propose we merge this and open separate PRs for supporting a configurable max data size, as already discussed in private with @stsquad.

CC @jiangliu @sboeuf

sboeuf commented on crates/devices/virtio-vsock/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+
+[dev-dependencies]
+vm-memory = { version = "0.7.0", features = ["backend-mmap", "backend-atomic"] }
+virtio-queue = { path = "../../virtio-queue", features = ["test-utils"] }
```
Ditto.

lauralt replied:
done

sboeuf reviewed (commented):
Looks good, but could you please provide more elaborate commit messages. I know you described the intent of this patchset through the PR description, but unfortunately it won't be part of the git history :)

lauralt commented:
> Looks good, but could you please provide more elaborate commit messages. I know you described the intent of this patchset through the PR description, but unfortunately it won't be part of the git history :)

Sure

lauralt commented on crates/devices/virtio-vsock/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+
+[dependencies]
+vm-memory = "0.7.0"
+virtio-queue = { path = "../../virtio-queue" }
```
Makes sense. I updated it.

lauralt commented:
@sboeuf I added more details to the commit messages. Please have another look.

lauralt commented on crates/devices/virtio-vsock/src/packet.rs:
```diff
@@ -0,0 +1,386 @@
+        &self.header
+    }
+
+    /// Return a reference to the `data` of the packet.
```
I'll address it in the next PR. I have to do some improvements to docs anyway. Thanks.
